### PR TITLE
Whitespace format

### PR DESCRIPTION
**Pull request by okomarov**
Whitespace formatting as per Egon's comments in https://github.com/matlab2tikz/matlab2tikz/issues/808.

Also, keeps crafted formatting intact. 

PS. I'll add H1 line and help at the end.

### CODE REVIEW

File: src/dev/formatWhitespace.m [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    % Check for mlint error
+    if ~isempty(tree.mtfind('Kind','ERR'))
+        [~,name,ext] = fileparts(obj.Filename);
+        error('formatWhitespace:mlintError','''%s'' contains syntax errors. Cannot proceed.',[name,ext])
+    end
+
+    % Smart indent except for crafted continuations of line
+    d.smartIndentContents()
+    lines      = textToLines(d.Text);
+    iDots      = ~cellfun('isempty',regexp(lines, '\.\.\.','once'));
```
Comment (egeerardyn) [inline src/dev/formatWhitespace.m]: Wouldn't this be more readable using just `strfind` (I thought there was also a MATLAB function to convert escape literal strings into regular expressions, but I can't find it back).

File: src/dev/formatWhitespace.m [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    import matlab.desktop.editor.*
+    try
+        d        = getActive();
+        oldLines = textToLines(d.Text);
+    catch
+        error('formatWhitespace:noEditorApi','Check that the Editor API is available.')
+    end
+
+    % Parse file into a tree
+    tree = mtree(d.Text);
```
Comment (egeerardyn) [inline src/dev/formatWhitespace.m]: I knew this kind of trickery existed somewhere in MATLAB. :smile: Is there actually some sort of documentation that you use for this? If so, I would really love to see it.

Comment (okomarov) [inline src/dev/formatWhitespace.m]: Yeah! It's called trial-and-error hahah (and reading the code...)

I used it in my https://github.com/okomarov/justify/blob/master/justify.m and you can check the use case.

File: src/dev/formatWhitespace.m (hunk 1 above)
Comment (okomarov) [inline src/dev/formatWhitespace.m]: strfind all the way :+1:

File: src/dev/formatWhitespace.m (hunk 2 above)
Comment (egeerardyn) [inline src/dev/formatWhitespace.m]: yay for trial-and-error (and undocumented MATLAB features)

**Comment by egeerardyn** [comment]
What I don't see in the code is how it is enforced that all functions are indented (i.e. how that setting is normalized across environments: that's a setting in MATLAB, but it is certainly not the default one).

Also, I think one of the things that may be worthwhile to implement as well, would be to specify a file name that needs to be processed (i.e. open it in the editor, process and close) and if none is supplied, just process the active editor.

**Comment by okomarov** [comment]
Function level indentation is enforced in the following way:

- Smart indent, i.e. reset indentation to 0 spaces.
- The last for loop adds 4 spaces to each block within a function. It does so for each function block in a sequential way, hence nested functions (that we do not have)  will have their content indented twice.

**Comment by egeerardyn** [comment]
Here it seems to have an unintended consequence that some levels are indented too much. I will produce an MWE where the problem shows together with my MATLAB settings. (edit will follow)

File: src/dev/formatWhitespace.m [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    % Find position of function/end pairs
+    tmp   = tree.mtfind('Kind','FUNCTION');
+    stpos = tmp.lineno;
+    enpos = tmp.lastone;
+
+    % Pad content with 4 spaces
+    for f = 1:numel(stpos)
+        pos        = stpos(f)+1:enpos(f)-1;
+        lines(pos) = strcat({'    '}, lines(pos));
```
Comment (egeerardyn) [inline src/dev/formatWhitespace.m]: Perhaps it would be better to define a variable `INDENT = repmat(' ',1,4);` at the beginning of the file.

**Comment by miscco** [comment]
Very nice, it was really confusing at the beginning, what indentation was intended.

**Comment by okomarov** [comment]
`MWEIndent.m` is indented correctly now.

File: src/dev/formatWhitespace.m [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        error('formatWhitespace:noEditorApi','Check that the Editor API is available.')
+    end
+
+    % Get specific document if filename is specified
+    if ~isempty(filename)
+        allDocs    = matlab.desktop.editor.getAll();
+        [~,fnames] = cellfun(@fileparts, {allDocs.Filename},'un',0);
+        [tf, pos]  = ismember(filename,fnames);
+        if ~tf
+            error('formatWhitespace:filenameNotFound','Filename "%s" not found.', filename)
```
Comment (egeerardyn) [inline src/dev/formatWhitespace.m]: Perhaps replace `tf` by a more descriptive name. Also, if I interpret this correctly, does this mean that this function only works with files that are currently open in the editor?
If so, please just add a `TODO: open file if it isn't open in the editor already`.

**Comment by egeerardyn** [comment]
:+1: Nice work! I think this will improve readability of our code base significantly.

When all current PRs are merged, we should create a PR with just indentation changed. That way, we avoid merge conflicts amongst different PRs.

**Comment by okomarov** [comment]
Ahem...here I already formatted m2t, as in  56ec8c0.

**Comment by egeerardyn** [comment]
Ah, that's why `matlab2tikz.m` was changed without any visible changes on GitHub :+1: I thought it was some weird kind of artefact from testing the procedure.

Anyhow, I'll try to do this regularly.

**Pull request merged**